cylc/flow/pathutil.py

```diff
@@ -335,15 +335,33 @@ def remove_empty_parents(
             break
 
 
-def get_next_rundir_number(run_path):
-    """Return the new run number"""
-    run_n_path = os.path.expanduser(os.path.join(run_path, "runN"))
-    try:
+def get_next_rundir_number(run_path: Union[str, Path]) -> int:
+    """Return the new run number
+
+    Args:
+        run_path: Top level installed workflow dir
+        (often ``~/cylc-run/workflow``).
+
+    """
+    run_n_path = Path(os.path.expanduser(os.path.join(run_path, "runN")))
+    if run_n_path.exists() and run_n_path.is_symlink():
         old_run_path = os.readlink(run_n_path)
-        last_run_num = re.search(r'(?:run)(\d*$)', old_run_path).group(1)
-        return int(last_run_num) + 1
-    except OSError:
-        return 1
+        # Line below could in theory not return a match group, so mypy objects.
+        # This function unlikely to be called in circumstances where this will
+        # be a problem.
+        last_run_num = re.search(                        # type: ignore
+            r'(?:run)(\d*$)', old_run_path).group(1)     # type: ignore
```

Comment:
Maybe error handling for if the `runN` symlink points to an invalid location would be nice, à la
https://github.com/cylc/cylc-flow/blob/7f635a285bee80210477ac3afb4c8f635083f581/cylc/flow/workflow_files.py#L1213-L1216

But it should ok to ignore this. Even better, ignore the specific rule only. Also, only needs to be on the first line.
```suggestion
        last_run_num = re.search(  # type: ignore[union-attr]
            r'(?:run)(\d*$)', old_run_path
        ).group(1)
```